File: cses/08_string_algorithms/1112_required-substring/solution.py
import sys

MOD = 10**9 + 7


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    pat = data[1] if len(data) > 1 else b""
    m = len(pat)

    # Mã hóa pattern về khoảng 0..25 (A = 0).
    P = [b - 65 for b in pat]

    # Prefix function (failure function) của KMP.
    pi = [0] * m
    for i in range(1, m):
        k = pi[i - 1]
        while k > 0 and P[i] != P[k]:
            k = pi[k - 1]
        if P[i] == P[k]:
            k += 1
        pi[i] = k

    # Xây KMP automaton: aut[j][c] = trạng thái kế tiếp từ trạng thái j khi đọc
    # ký tự c. j chạy 0..m-1 (chưa khớp hết); nếu aut[j][c] == m tức pattern
    # đã xuất hiện đầy đủ.
    aut = [[0] * 26 for _ in range(m)]
    for j in range(m):
        pj = P[j]
        row = aut[j]
        if j == 0:
            for c in range(26):
                row[c] = 1 if c == pj else 0
        else:
            # Không khớp thì nhảy theo failure link pi[j-1].
            prev = aut[pi[j - 1]]
            for c in range(26):
                row[c] = j + 1 if c == pj else prev[c]

    # Nén chuyển tiếp: với mỗi trạng thái j, gom các ký tự dẫn tới cùng một
    # trạng thái < m (bỏ ký tự dẫn tới m vì các xâu đó đã chứa pattern).
    trans = []
    for j in range(m):
        cnt = {}
        row = aut[j]
        for c in range(26):
            nx = row[c]
            if nx < m:
                cnt[nx] = cnt.get(nx, 0) + 1
        trans.append(list(cnt.items()))

    # DP: dp[j] = số xâu độ dài hiện tại, chưa bao giờ đạt trạng thái m, kết
    # thúc ở trạng thái j. Khởi tạo dp[0] = 1 cho xâu rỗng.
    dp = [0] * m
    dp[0] = 1
    for _ in range(n):
        ndp = [0] * m
        for j in range(m):
            v = dp[j]
            if v:
                for nx, k in trans[j]:
                    ndp[nx] = (ndp[nx] + v * k) % MOD
        dp = ndp

    # Tổng dp là số xâu không chứa pattern; kết quả = 26^n trừ đi số đó.
    not_contain = sum(dp) % MOD
    ans = (pow(26, n, MOD) - not_contain) % MOD
    sys.stdout.write(str(ans) + "\n")


main()

File: cses/08_string_algorithms/1112_required-substring/solution.cpp
#include <algorithm>
#include <array>
#include <iostream>
#include <string>
#include <vector>

namespace {
constexpr long long MOD = 1'000'000'007LL;

// Lũy thừa nhanh theo modulo, dùng để tính 26^n.
long long modular_power(long long base, int exponent) {
    long long result = 1;
    while (exponent > 0) {
        if ((exponent & 1) != 0) {
            result = result * base % MOD;
        }
        base = base * base % MOD;
        exponent >>= 1;
    }
    return result;
}
}  // namespace

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    std::string pattern;
    std::cin >> n >> pattern;
    const int m = static_cast<int>(pattern.size());

    // Prefix function (failure function) của KMP trên pattern.
    std::vector<int> prefix(static_cast<std::size_t>(m), 0);
    for (int i = 1; i < m; ++i) {
        int matched = prefix[static_cast<std::size_t>(i - 1)];
        while (matched > 0 &&
               pattern[static_cast<std::size_t>(i)] != pattern[static_cast<std::size_t>(matched)]) {
            matched = prefix[static_cast<std::size_t>(matched - 1)];
        }
        if (pattern[static_cast<std::size_t>(i)] ==
            pattern[static_cast<std::size_t>(matched)]) {
            ++matched;
        }
        prefix[static_cast<std::size_t>(i)] = matched;
    }

    // Xây KMP automaton: automaton[state][letter] = trạng thái kế tiếp khi
    // đang ở trạng thái state và đọc ký tự letter. Nếu bằng m tức pattern
    // vừa xuất hiện đầy đủ.
    std::vector<std::array<int, 26>> automaton(static_cast<std::size_t>(m));
    for (int state = 0; state < m; ++state) {
        for (int letter = 0; letter < 26; ++letter) {
            const char current = static_cast<char>('A' + letter);
            if (current == pattern[static_cast<std::size_t>(state)]) {
                // Ký tự khớp: tiến sang trạng thái state + 1.
                automaton[static_cast<std::size_t>(state)][static_cast<std::size_t>(letter)] =
                    state + 1;
            } else if (state == 0) {
                automaton[0][static_cast<std::size_t>(letter)] = 0;
            } else {
                // Không khớp: nhảy theo failure link để tận dụng phần đã khớp.
                automaton[static_cast<std::size_t>(state)][static_cast<std::size_t>(letter)] =
                    automaton[static_cast<std::size_t>(prefix[static_cast<std::size_t>(state - 1)])]
                             [static_cast<std::size_t>(letter)];
            }
        }
    }

    // DP đếm số xâu KHÔNG chứa pattern: dp[state] = số xâu đang xét chưa bao
    // giờ chạm trạng thái m và hiện dừng ở trạng thái state.
    std::vector<long long> dp(static_cast<std::size_t>(m), 0);
    std::vector<long long> next(static_cast<std::size_t>(m), 0);
    dp[0] = 1;  // Xâu rỗng ở trạng thái 0.
    for (int length = 0; length < n; ++length) {
        std::fill(next.begin(), next.end(), 0);
        for (int state = 0; state < m; ++state) {
            const long long ways = dp[static_cast<std::size_t>(state)];
            if (ways == 0) {
                continue;
            }
            // Thử thêm cả 26 ký tự; chỉ giữ chuyển tiếp có trạng thái < m
            // (bỏ trạng thái m vì khi đó xâu đã chứa pattern).
            for (int letter = 0; letter < 26; ++letter) {
                const int new_state =
                    automaton[static_cast<std::size_t>(state)][static_cast<std::size_t>(letter)];
                if (new_state < m) {
                    long long& destination = next[static_cast<std::size_t>(new_state)];
                    destination += ways;
                    if (destination >= MOD) {
                        destination -= MOD;
                    }
                }
            }
        }
        dp.swap(next);
    }

    // Tổng dp là số xâu không chứa pattern; kết quả = 26^n trừ đi số đó.
    long long avoiding = 0;
    for (const long long ways : dp) {
        avoiding += ways;
        if (avoiding >= MOD) {
            avoiding -= MOD;
        }
    }
    const long long answer = (modular_power(26, n) - avoiding + MOD) % MOD;
    std::cout << answer << '\n';
    return 0;
}
